consumer/src/main/java/com/flipkart/varadhi/consumer/delivery/MessageDelivery.java

```diff
@@ -0,0 +1,71 @@
+package com.flipkart.varadhi.consumer.delivery;
+
+import com.flipkart.varadhi.entities.Endpoint;
+import com.flipkart.varadhi.entities.Message;
+import com.flipkart.varadhi.exceptions.NotImplementedException;
+import com.google.common.collect.Multimap;
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+
+public interface MessageDelivery {
+    static MessageDelivery of(Endpoint endpoint) {
+        return switch (endpoint.getProtocol()) {
+            case HTTP1_1 -> new HttpMessageDelivery(endpoint);
+            case HTTP2 -> throw new NotImplementedException("HTTP2 is not supported yet");
+            default -> throw new IllegalArgumentException("Unsupported protocol: " + endpoint.getProtocol());
+        };
+    }
+
+    CompletableFuture<DeliveryResponse> deliver(Message message)
+            throws Exception;
+
+    class HttpMessageDelivery implements MessageDelivery {
+        private final Endpoint.HttpEndpoint endpoint;
+        private final HttpClient httpClient;
+
+        public HttpMessageDelivery(Endpoint endpoint) {
+            this.endpoint = (Endpoint.HttpEndpoint) endpoint;
+            this.httpClient = HttpClient.newBuilder()
+                    .version(this.endpoint.isHttp2Supported() ? HttpClient.Version.HTTP_2 : HttpClient.Version.HTTP_1_1)
+                    .connectTimeout(Duration.ofMillis(this.endpoint.getConnectTimeoutMs()))
+                    .followRedirects(HttpClient.Redirect.NORMAL)
+                    .build();
+        }
+
+        @Override
+        public CompletableFuture<DeliveryResponse> deliver(Message message) throws Exception {
+            HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
+                    .uri(endpoint.getUrl().toURI())
+                    .timeout(Duration.ofMillis(endpoint.getRequestTimeoutMs()))
+                    .header("Content-Type", endpoint.getContentType())
+                    .method(
+                            endpoint.getMethod(),
+                            ArrayUtils.isEmpty(message.getPayload()) ? HttpRequest.BodyPublishers.noBody() :
+                                    HttpRequest.BodyPublishers.ofByteArray(message.getPayload())
```

Comment:
thing to note here: 
`HttpRequest.BodyPublishers.ofByteArray` does not make a copy of the array we are passing, which is good. But turns out that httpClient from jdk, does not support ByteBuffer object directly (and does not support off-heap objects as well.)

If that is what we ned in future, then we will want to look into using netty directly or an http client that does support offheap buffers. I wonder if vertx's http client supports it or not!!

Reply:
<img width="596" alt="Screenshot 2024-06-18 at 8 45 39 AM" src="https://github.com/flipkart-incubator/varadhi/assets/391976/4675b9e4-4908-4336-ae99-42a74e4c4004">

Interesting. Something to keep in mind. For now jdk's http client is fine.